utils/gridlink_impl.c.src

```diff
@@ -179,113 +142,163 @@ cellarray_DOUBLE * gridlink_DOUBLE(const int64_t NPART,
         if (ix>nmesh_x-1)  ix--;    /* this shouldn't happen, but . . . */
         if (iy>nmesh_y-1)  iy--;
         if (iz>nmesh_z-1)  iz--;
-        XRETURN(X[i] >= xmin && X[i] <= xmax, NULL,
-               "x[%"PRId64"] = %"REAL_FORMAT" must be within [%"REAL_FORMAT",%"REAL_FORMAT"]\n",
-                i, X[i], xmin, xmax);
-        XRETURN(Y[i] >= ymin && Y[i] <= ymax, NULL,
-               "y[%"PRId64"] = %"REAL_FORMAT" must be within [%"REAL_FORMAT",%"REAL_FORMAT"]\n",
-               i, Y[i], ymin, ymax);
-        XRETURN(Z[i] >= zmin && Z[i] <= zmax, NULL,
-               "z[%"PRId64"] = %"REAL_FORMAT" must be within [%"REAL_FORMAT",%"REAL_FORMAT"]\n",
-               i, Z[i], zmin, zmax);
-
-        XRETURN(ix >= 0 && ix < nmesh_x, NULL, "ix=%d must be within [0,%d)\n", ix, nmesh_x);
-        XRETURN(iy >= 0 && iy < nmesh_y, NULL, "iy=%d must be within [0,%d)\n", iy, nmesh_y);
-        XRETURN(iz >= 0 && iz < nmesh_z, NULL, "iz=%d must be within [0,%d)\n", iz, nmesh_z);
+    
+        out_of_bounds += ix < 0 || ix >= nmesh_x || iy < 0 || iy >= nmesh_y || iz < 0 || iz >= nmesh_z;
 
         const int64_t icell = CONVERT_3D_INDEX_TO_LINEAR(ix, iy, iz, nmesh_x, nmesh_y, nmesh_z);
-        if(options->copy_particles == 0) {
-            all_cell_indices[i] = icell;
+        all_cell_indices[i] = icell;
+        if(!options->copy_particles){
             original_indices[i] = i;
+        }
+    }
+    
+    XRETURN(out_of_bounds == 0, NULL, "%"PRId64" particles are out of bounds. Check periodic wrapping?\n", out_of_bounds);
+    
+    // Now determine the number of particles per cell so each cell knows its global starting index
+    // This information is used for both the copy and in-place versions
+#if defined(_OPENMP)
+    int max_nthreads = omp_get_max_threads();
```

Comment:
Huh - do you need `max_nthreads` here? Can you not directly use `nthreads` through `omp_get_num_threads()`?

Reply:
Yes, this actually needs `max_threads`!  This is a serial region, so `omp_get_num_threads()` will always return 1.

Reply:
Hmm - in that case, we should set `max_threads` to `num_threads` as well (right after we set `num_threads` with `omp_set_num_threads`)

Reply:
Okay, I've just renamed it `nthreads` and assigned it the "real" value upon entry into the first parallel region.  Will be the same in most cases.